client/src/pages/SearchBar.tsx

```diff
@@ -0,0 +1,101 @@
+import { useEffect, useState } from 'react';
+import { type Product } from './Catalog';
+import { ProductCard } from './Catalog';
+
+export type Props = {
+  toggleSearch: () => void;
+};
+
+export function SearchBar({ toggleSearch }: Props) {
+  const [searchText, setSearchText] = useState('');
+  const [searchResults, setSearchResults] = useState<Product[]>([]);
+  const [error, setError] = useState<unknown>();
+  const [isLoading, setIsLoading] = useState(false);
+
+  useEffect(() => {
+    if (searchText.length < 2) {
+      setSearchResults([]);
+      return;
+    }
+
+    const timerId = setTimeout(() => {
```

Comment:
Sending search queries to the backend only after a pause is called "debouncing". The approach you use here is somewhat simplistic and doesn't give the best results possible. Some time in the future you could play with improving it. Here is an article that could give you a good solution:
https://medium.com/nerd-for-tech/debounce-your-search-react-input-optimization-fd270a8042b